txpool/txpool.go

```diff
@@ -285,6 +272,44 @@ func (t *TxPool) validateTx(tx *types.Transaction) error {
 	if tx.Gas < intrinsicGas {
 		return ErrIntrinsicGas
 	}
+
+	// Check if the transaction has a strictly positive value
+	if tx.Value.Sign() < 0 {
+		return ErrNegativeValue
+	}
+
+	if !t.dev && tx.From != types.ZeroAddress {
+		// Only if we are in dev mode we can accept
+		// a transaction without validation
+		return ErrNonEncryptedTxn
+	}
+
+	// Check if the transaction is signed properly
+	var signerErr error
+	if tx.From == types.ZeroAddress {
+		tx.From, signerErr = t.signer.Sender(tx)
+		if signerErr != nil {
+			return ErrInvalidSender
+		}
+	}
+
+	// Grab the state root for the latest block
+	stateRoot := t.store.Header().StateRoot
+	accountBalance := t.store.GetBalance(stateRoot, tx.From)
+
+	// Check if the sender has enough funds to execute the transaction
+	if accountBalance.Cmp(tx.Cost()) < 0 {
+		return ErrInsufficientFunds
+	}
+
+	// Check nonce ordering
+	if t.store.GetNonce(stateRoot, tx.From) > tx.Nonce {
+		return ErrNonceTooLow
+	}
```

Comment:
Maybe do nonce check first, it should be faster
```suggestion
	// Grab the state root for the latest block
	stateRoot := t.store.Header().StateRoot
	// Check nonce ordering
	if t.store.GetNonce(stateRoot, tx.From) > tx.Nonce {
		return ErrNonceTooLow
	}

	// Check if the sender has enough funds to execute the transaction
	accountBalance := t.store.GetBalance(stateRoot, tx.From)
	if accountBalance.Cmp(tx.Cost()) < 0 {
		return ErrInsufficientFunds
	}
```

Reply:
Fixed 👍 

74b6cc3

Reply:
I see that you've moved it but the crucial part is not moved.
You still call `t.store.GetBalance(stateRoot, tx.From)` before checking the `nonce`.
If nonce is invalid you have fetched balance for nothing.
Please move that part after the nonce ordering check.

Reply:
I thought I've moved it properly, but I've only moved out part of it.

Sorry about that 👍 

Fixed 👍 

4423ebb